### ISSUE TITLE
Tag editor shortcomings?

### ISSUE THREAD

**Issue by 0xdabbad00** [FSecureLABS/awspx#1]
Author of CloudMapper here. First, awesome work!  I just started looking at your project and the code looks very clean (especially in comparison to many of the other AWS security project code bases, ex. CloudMapper :) ).  In your post you mentioned the use of AWS tag editor and that there were short-comings with that.  I haven't seen that approach used before, and was surprised to not find a hundred List and Describe calls as is seen with other projects that try to identify all of the resources in an account.  Where are you actually making the calls to get the resources for the account?  I saw https://github.com/FSecureLABS/awspx/blob/136e64a2dc2f4abd931907dcbafe4b14663d4c64/cli.py#L235 which I believe then calls into this https://github.com/FSecureLABS/awspx/blob/136e64a2dc2f4abd931907dcbafe4b14663d4c64/lib/aws/ingestor.py#L802
But I wasn't able to follow how that collection actually happens?  Also, what were the short-comings you ran into?

Also, there are some projects that have some similarities with this one that you might be interested in.
- Lyft's `cartography` has explored trying to collect all the resources in an account graph and make those queryable. You can see for example here how it does some parsing of resource policies on S3 buckets: https://github.com/lyft/cartography#which-s3-buckets-have-a-policy-granting-any-level-of-anonymous-access-to-the-bucket
- Duo's `parliament` (which I developed) parses the AWS docs for information about IAM privileges: https://github.com/duo-labs/parliament/blob/master/utils/update_iam_data.py . This generates a json file similar to your own https://github.com/FSecureLABS/awspx/blob/master/lib/aws/actions.py
- Duo's `CloudMapper` (which I also authored and maintain) has two things that might be of interest.  First, is the ability to generate a graph of IAM relationships (users, roles, policies, groups).  It is very basic as it doesn't try to understand the policies at all, just shows which principals have which policies. Some info about that functionality is at https://summitroute.com/blog/2019/03/11/cloudmapper_iam_report_command/ and the relevant screenshot included.

<img width="583" alt="iam_report_graph" src="https://user-images.githubusercontent.com/1979334/70248798-b590e200-1738-11ea-913d-73facfef6c7a.png">

The second feature of CloudMapper I want to point out that is likely going to be more interesting to you is the recently added `access_check` command (recent as in I only added it a week ago).  This uses `parliament` to parse the policies and understand what types of resources different privileges impact (ex. it automatically figures out that s3:GetObject is used on resources with an ARN of an S3 object).  Then it tries to actually understand the policies by combining them all together, so if a user has been granted `s3:GetObject` on `*`, but also has a `Deny` `*` on `*`, then it understands that the user has no privileges.  I've barely been able to document it, but I discussed it a little at https://github.com/duo-labs/cloudmapper/releases/tag/2.8.1

Anyway, looking forward to hearing about how you used the tag editor API and what those short-comings were.

**Comment by GeneralTesler** [comment]
Not the author, but boto3 allows you to do some cool stuff with API enumeration and listing resources. For example, if you want to list EC2 instance API calls, you can do 

```
resource = session.resource("ec2")
operations = resource.instances._handler.service_context.service_model.operation_names
[print(op) for op in operations]
```

And if you want to list out IAM users, you can do

```
resource = session.resource("iam")
[print(r.arn) for r in getattr(resource, "users").all()]
```
which calls the ListUsers API with a JMESPath of Users[] on the results.

As far as I can tell, this is essentially using this method to list resources. E.g. it gets the resource (EC2), pulls out all of the sub-resources (DHCP options, IP addresses, etc), then calls the listing handlers. https://github.com/FSecureLABS/awspx/blob/136e64a2dc2f4abd931907dcbafe4b14663d4c64/lib/aws/ingestor.py#L391 and https://github.com/FSecureLABS/awspx/blob/136e64a2dc2f4abd931907dcbafe4b14663d4c64/lib/aws/ingestor.py#L114 seem to be the most relevant

**Comment by beatro0t** [comment]
Great to meet you @0xdabbad00 . I referenced you in the post - on behalf of the world, thank you for  taking the time to try and work out how many actions there actually are! :) Also, great work with Cloudmapper, I've used it myself to explore the networking side of AWS in the past, and found it to be very useful.

So the tag editor does not have an API (at least when I last looked), which means you first have to download a CSV and then parse it, which sucks. Secondly, it is very limited with the information it provides; it doesn't include _Arn_ information, so this is something you would need to construct yourself by using the _name_, _region_, _resource type_, etc. 

In this project's earlier days, this is how we did things. The problem is that if you want any additional info (which is useful from a front-end perspective) then you will need to go the boto route, making the whole exercise kind of pointless.

Also, I'm pretty sure the tag editor is only viable for resources that can be tagged, so there's probably a bunch you'd miss (just double check because the last time I looked was a year ago and "what you can tag" has been extended since). Something that might be worth checking out is the AWS config service (thought it would be a bit too invasive for this project but if you want something more continuous its probably a good option).

As for the ingestor, you're absolutely right. When I initially added the resource collection logic it was -_Get_, _List_, _Describe_, etc but @masteryeets recently found a clever way to use boto's own iterators - as @GeneralTesler  described perfectly! :) - which let us replace a ton of that sort of code. Basically cli.py imports '*' from lib/aws/ingestor.py which contains a bunch of classes inheriting from the class 'Ingestor'. This base class is where this boto logic sits and the other classes (e.g. S3, EC2, etc) inherit from it. The functions in those classes (eg. EC2) are actually just additional functions that augment what we couldn't get generically from boto. Check out [CONTRIBUTING.md](https://github.com/FSecureLABS/awspx/blob/master/CONTRIBUTING.md) for more info, but the gist of it is that we wanted to make it easy to add support for new services, which you can do in a rudimentary fashion with as little code as:
```
class CloudFormation:
    pass
```
Thanks for the list - I had a look at Cartography a while back and think its a great project (with an awesome name)! They have some fantastic ideas and it looks incredibly promising. First time hearing about parliament though (which is frustrating because it's precisely what I was looking for not even a few weeks ago). I just had a look and I really like how you've crawled and scraped for services; we defined them explicitly, which as I'm sure you can imagine is tedious. If there's anything you can use from update_actions.py, please take it - I'd far rather refresh actions using something someone else is maintaining :)

In regard to graphing principals and policies - awesome. I know first hand how much work it is to gather and model that sort of information. If you ever find that there are limitations in what you can do from the awspx front-end, I recommend using neo4j's visualizer instead - which is available on localhost:7474 (neo4j:neo4j). The DB is modeled quite sensibly - I'm obviously biased ;) - so its pretty straight-forward to achieve the same thing using cypher:

``` 
MATCH attached_managed_policies=()-[:TRANSITIVE*1..]-(:`AWS::Iam::Policy`) RETURN attached_managed_policies 
```

The added benefit is that if you're dealing a particularly large environment, you can write refined queries or return text instead.

Cloudmapper's _access_check_ sounds epic and I'm very keen to check it out. Its something I looked into doing but it becomes quite complicated. Preproccessing it would mean losing _Deny_ information which is important for computing attacks. For example, its pretty interesting to see what privileges could be gained by detaching policies imposing restrictions. If you don't know what those restrictions are anymore, you can't tell what you've gained. This is not a case we've added to the attack computation logic but its something we'd like to be able to cater to in the future.

That being said, there are many benefits that come with being able to consolidate the overall effect actions have. For one, it makes the a graph a lot more readable (by removing information that is for the most-part redundant), and two, I find its incredibly irritating when you see **All AWS Accounts** available to you in the awspx search bar, but find it only exists due to a resource-based policy or something denying public access.

As an aside, something you might be interested in (which we didn't document at all) are associative relationships (basically the dotted edges that sometimes appear when you double click, eg on an EC2 instance). If you run the following command in the neo4j visualizer:

```  
MATCH network=(a:`AWS::Ec2::Instance`:Resource)-[:ASSOCIATIVE*0..]-(:`AWS::Ec2::Instance`:Resource) RETURN network  
```


You will get all EC2 instances, and all other EC2 instances they can potentially communicate with, due to the relationships that naturally exist between these resources. If you restrict the result set further by taking into account the resources on the path (eg, security groups, nacls, routing tables,  etc), you can get really accurate results (since you know exactly what to expect, you can process them in the right way). This is not something I have looked into properly but imagine all the right information to be there for it to be feasible.

I hope that answers your questions.

**Comment by 0xdabbad00** [comment]
Thank you for the detailed response.  I was very interested when I saw how you identify some of the relationships between resources. One of the problems I ran into with `access_check` was if I want to find out who can impact an EC2, I also usually want to know who can impact things directly related to that resource, such as the VPC it is in, Security Groups associated with it, etc.  which I saw you've identified at https://github.com/FSecureLABS/awspx/blob/master/lib/aws/ingestor.py#L777

As for your last paragraph about identifying EC2 instances that an EC2 can communicate with, that was the original purpose of CloudMapper.  The lines shown below between resources are the communication paths that have been identified based on Security Groups.  I only consider Security Groups though and not NACLs or routing tables.  However, it should be fairly complete and can map out much more than EC2s, such as ECS, RDS, Redshift, etc.

![image](https://user-images.githubusercontent.com/1979334/70330732-5f38a780-17fb-11ea-87a6-099f599194e9.png)

Anyway, thank you for answering my question.  I'll have to dig into that technique of using boto.
